arrow2_convert/src/deserialize.rs

```diff
@@ -48,23 +48,36 @@ pub trait ArrowDeserialize: ArrowField + Sized {
     /// something like for<'a> &'a T::ArrayType: IntoIterator<Item=Option<E>>,
     /// However, the E parameter seems to confuse the borrow checker if it's a reference.
     fn arrow_deserialize_internal(
-        v: <<Self::ArrayType as IterRef>::Iter<'_> as Iterator>::Item,
+        v: <<Self::ArrayType as RefIntoIterator>::Iterator<'_> as Iterator>::Item,
     ) -> <Self as ArrowField>::Type {
         Self::arrow_deserialize(v).unwrap()
     }
-}
-
-/// Internal trait used to support deserialization and iteration of structs, and nested struct lists
-///
-/// Trivial pass-thru implementations are provided for arrow2 arrays that auto-implement IterRef.
-///
-/// The derive macro generates implementations for typed struct arrays.
-#[doc(hidden)]
-pub trait ArrowArray: IterRef {
-    type BaseArrayType: Array;
 
-    // Returns a typed iterator to the underlying elements of the array from an untyped Array reference.
-    fn iter_from_array_ref(b: &dyn Array) -> <Self as IterRef>::Iter<'_>;
+    #[inline]
+    #[doc(hidden)]
+    /// For internal use only
+    ///
+    /// TODO: this can be removed up by using arrow2::array::StructArray and
+    /// arrow2::array::UnionArray to perform the iteration for unions and structs
+    /// which should be possible if structs and unions are deserialized via scalars.
+    ///
+    /// Helper to return an iterator for elements from a [`arrow2::array::Array`].
+    ///
+    /// Overridden by struct and enum arrays generated by the derive macro, to
+    /// downcast to the arrow2 array type.
+    fn arrow_array_ref_into_iter(
+        array: &dyn Array,
+    ) -> Option<<Self::ArrayType as RefIntoIterator>::Iterator<'_>>
+    where
+        Self::ArrayType: 'static,
```

Comment:
Does this have to be 'static? Or just within the lifetime of &dyn?

Reply:
I tried with a named lifetime. I'm not fully sure, but I think `'static` is needed for the `downcast_ref` to facilitate the vtable generation.

Reply:
Ohh, right, because `Any: 'static` bound. Nevermind then!

Might even add a comment to avoid future confusion if you'd like:

```
// because of `Any: 'static` bound
```